designs/2020-es-module-support/README.md

```diff
@@ -0,0 +1,156 @@
+
+### Nomenclature
+
+* `ESM` ecmascript module. The standardized module format for javascript
+* `CJS` commonjs. Node's historical module format (i.e., `module.exports = ...`, `const a = require('foo')`)
+* `UMD` universal module definition. A legacy format for specifying javascript code as a module in a way that tries to guarantee compatibility with many loader formats including requireJS, commonjs, and global script tags.
+* `IIFE` immediately invoked function expressions. A legacy strategy for encapsulating javascript into a module that won't leak global variables into the execution environment. It is still the primary way of packaging global script tags for inclusion in a web page. `UMD` is usually wrapped in an `IIFE`
+* `CDN` content delivery network. can be thought of as a set of servers that host static content all around the globe to deliver assets quickly and at scale.
+
+
+### build process
+
+Today, espree is written in commonjs (CJS) format, and a universal module definition (UMD) build is created too.
+
+
+today's build process:
+```
+┌-------------------┐               ┌-----------------┐
+│                   │               │                 │                package.json:
+│     espree.js     |               │ build/espree.js │
+│ (CJS entry point) ├──BUILD_STEP──▶│   (UMD bundle)  │     CJS ---▶   "main": "espress.js",
+│                   │               │                 │
+└-------------------┘               └-----------------┘
+```
+
+
+The strategy that probably makes the most sense in this initial work is to do what is known as a "dual package" solution, where
+espree provides an ESM, and a CJS module.
+
+Providing a dual package shouldn't affect any existing espree users, but provies an ESM option for users that want it.
+
+
+proposed build process in this RFC:
+```
+                                  ┌-------------------┐
+                                  │                   │
+                                  │  dist/espree.cjs  │
+                                  │ (CJS entry point) │                package.json:
+┌-------------------┐             │                   │
+│                   │             └--▲----------------┘     CJS ---▶   "main": "dist/espress.cjs",
+│     espree.js     │                │                                 "exports": {
+│ (ESM entry point) ├───BUILD_STEP───┘                      ESM ---▶     "import": "espree.js",
+│                   │                                       CJS ---▶     "require": "./dist/espree.cjs"
+└-------------------┘                                                  },
+                                                                       "type": "module"
+```
+
+browserify is specifically oriented around commonjs as the input format, so I also propose replacing it with rollup, which understands and expects ESM as it's input format.
+
+`UMD` is not part of the formal package interface and it's usage is questionable, so I propose we drop it altogether.
```

Comment:
As mentioned by @nzakas in https://github.com/eslint/rfcs/pull/72/files#r548708551, we just use browserify/UMD for https://eslint.org/demo. Once this change is made, we still need _some_ way to load espree in the browser, but we can use rollup or whatever else is easiest since we have no API compatibility constraints there. So I’m :+1: to getting rid of UMD if there’s a better option for the browser demo.

Reply:
I think the current build system should still work fine, but we can double check.

Reply:
For the ESLint demo https://eslint.org/demo we're now using [webpack](https://github.com/eslint/website/blob/master/webpack.config.js).

Espree demo was dropped in favor of https://astexplorer.net/

I couldn't find that we are now using the browserify script nor the bundle it creates for anything, so made a PR https://github.com/eslint/espree/pull/466 to remove it.

Reply:
for demo pages, I'd say just go crazy with it and use the browser's built-in module support. e.g., this works today:

```html
<script type="module">
import espree from 'https://cdn.skypack.dev/espree'

// use espree normally here
</script>
```

🤯